File: 213-house-robber-ii/house-robber-ii.py
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        
        def rec(index, end):
            if index >= end:
                return 0
            if dp[index] != -1:
                return dp[index]
            
            # Choose to rob the current house
            pick = nums[index] + rec(index + 2, end)
            # Skip the current house
            notPick = rec(index + 1, end)
            
            dp[index] = max(pick, notPick)
            return dp[index]

        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums)
        if len(nums) == 3:
            return max(nums)

        # First case: Rob houses from index [0, n-2]
        dp = [-1] * len(nums)
        case1 = rec(0, len(nums) - 1)

        # Second case: Rob houses from index [1, n-1]
        dp = [-1] * len(nums)
        case2 = rec(1, len(nums))

        return max(case1, case2)

File: 213-house-robber-ii/house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        if(nums.length == 1){
            return nums[0];
        }
        int [] dp = new int[nums.length + 1];
        Arrays.fill(dp,-1);
        int yes = fun(0,1,nums,dp);
        Arrays.fill(dp,-1);
        int no = fun(1,0,nums,dp);
        return Math.max(yes,no);
    }
    public int fun(int index, int flag,int [] nums,int [] dp){
        if(index >= nums.length){
            return 0;
        }
        if(dp[index] != -1){
            return dp[index];
        }
        int pick = 0;
        if(flag == 1){
            if(index < nums.length-1){
                pick = fun(index + 2,flag,nums,dp) + nums[index];
            }
        }
        else{
            pick = fun(index + 2,flag,nums,dp) + nums[index];
        }
        int notPick = fun(index + 1,flag,nums,dp);
        dp[index] = Math.max(pick, notPick);
        return dp[index];
    }
}